File: Leetcode/FindUniqueBinaryString.rs
// https://leetcode.com/problems/find-unique-binary-string

struct Solution {}

impl Solution {
    pub fn find_different_binary_string(nums: Vec<String>) -> String {
        let n = nums.len();
        let mut x = vec![false; (1 << 16) + 1];

        for num in &nums {
            let mut p = 1;
            let mut a = 0;
            for y in num.chars().rev() {
                a |= ((y == '1') as usize) * p;
                p <<= 1;
            }
            x[a] = true;
        }

        let ans = x.iter().position(|x| x == &false).unwrap();
        let y = format!("{:#018b}", ans);

        let y = y.chars().rev().collect::<String>();
        let y = y.chars().take(n).collect::<String>();
        y.chars().rev().collect::<String>()
    }
}

File: Leetcode/FindUniqueBinaryString.cpp
// https://leetcode.com/problems/find-unique-binary-string

class Solution {
   public:
    string findDifferentBinaryString(vector<string>& nums) {
        sort(nums.begin(), nums.end());
        nums.push_back("");

        auto inc = [](string& ans) {
            const auto n = ans.length();

            int carry = 1;
            for (int i = n - 1; i >= 0; --i) {
                ans[i] += carry;
                carry = 0;
                if (ans[i] > '1') {
                    ans[i] = '0';
                    carry = 1;
                }
            }
        };

        string cur(nums[0].size(), '0');
        for (size_t i = 0; i < nums.size();) {
            if (nums[i] == cur) {
                inc(cur);
                ++i;
            } else {
                return cur;
            }
        }

        unreachable();
    }
};